pkg/collector/reconcile/servicemonitor.go

```diff
@@ -0,0 +1,170 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/open-telemetry/opentelemetry-operator/pkg/featuregate"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
+)
+
+// +kubebuilder:rbac:groups=monitoring.coreos.com,resources=servicemonitors,verbs=get;list;watch;create;update;patch;delete
+
+// ServiceMonitors reconciles the service monitor(s) required for the instance in the current context.
+func ServiceMonitors(ctx context.Context, params Params) error {
+	if !params.Instance.Spec.Observability.Metrics.CreateServiceMonitors || !featuregate.PrometheusOperatorIsAvailable.IsEnabled() {
+		return nil
+	}
+
+	desired := desiredServiceMonitors(ctx, params)
+
+	// first, handle the create/update parts
+	if err := expectedServiceMonitors(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the expected service monitors: %w", err)
+	}
+
+	// then, delete the extra objects
+	if err := deleteServiceMonitors(ctx, params, desired); err != nil {
+		return fmt.Errorf("failed to reconcile the service monitors to be deleted: %w", err)
+	}
+
+	return nil
+}
+
+func desiredServiceMonitors(_ context.Context, params Params) []monitoringv1.ServiceMonitor {
+	col := params.Instance
+	return []monitoringv1.ServiceMonitor{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: col.Namespace,
+				Name:      naming.ServiceMonitor(col),
+				Labels: map[string]string{
+					"app.kubernetes.io/name":       naming.ServiceMonitor(params.Instance),
+					"app.kubernetes.io/instance":   fmt.Sprintf("%s.%s", params.Instance.Namespace, params.Instance.Name),
+					"app.kubernetes.io/managed-by": "opentelemetry-operator",
+				},
+			},
+			Spec: monitoringv1.ServiceMonitorSpec{
+				Endpoints: []monitoringv1.Endpoint{{
+					Port: "monitoring",
```

Comment:
Where is this port being defined?  Does this handle situations where someone manually configures the collector's metrics port to be something other than the default?

Reply:
It is defined here:
https://github.com/open-telemetry/opentelemetry-operator/blob/56f5de94dd3f1f83ac7dbdba3b8b4a6379317b48/pkg/collector/reconcile/service.go#L197-L200

It cannot be modified (right now). But I think it could be an interesting follow-up task: to allow the personalization of the metrics port. What do you t hink?

Reply:
Ok, I see your point.
Yes, the metrics become unreachable. If we want to address that use case, I would do it as part of a different issue since it would involve changes in https://github.com/open-telemetry/opentelemetry-operator/blob/56f5de94dd3f1f83ac7dbdba3b8b4a6379317b48/pkg/collector/reconcile/service.go#L197-L200

And look for a way to get the address where the metrics are exposed. What do you think?

Reply:
Doing it as part of another issue is good.  I kinda feel like we need to solve that issue before we introduce this functionality.

Reply:
Created #1931